File: src/services/shiftService.ts
import {
  collection,
  addDoc,
  updateDoc,
  deleteDoc,
  doc,
  getDoc,
  query,
  where,
  getDocs,
} from 'firebase/firestore';
import { db } from './firebase';
import { auth } from './firebase';
import { Shift, ShiftType, ShiftStatus, SHIFT_CONFIG, SHIFT_RULES } from '../types';
import { getWeekStart, getWeekEnd } from '../utils/dateUtils';

async function createNotification(
  userId: string,
  type: 'shift_approved' | 'shift_rejected' | 'hours_limit',
  title: string,
  body: string
): Promise<void> {
  try {
    await addDoc(collection(db, 'notifications'), {
      userId,
      type,
      title,
      body,
      read: false,
      createdAt: new Date().toISOString(),
    });
  } catch {
    // notification failure is non-critical
  }
}

async function syncEmployeeWeeklyHours(employeeId: string, shiftDate: string): Promise<void> {
  try {
    const weekDate = new Date(shiftDate);
    const weekStart = getWeekStart(weekDate);
    const weekEnd = getWeekEnd(weekDate);
    const q = query(
      collection(db, 'shifts'),
      where('employeeId', '==', employeeId),
      where('status', '==', 'approved'),
      where('date', '>=', weekStart),
      where('date', '<=', weekEnd)
    );
    const snap = await getDocs(q);
    const totalHours = snap.docs
      .map(d => d.data() as Shift)
      .reduce((sum, s) => sum + s.durationHours, 0);
    const userSnap = await getDoc(doc(db, 'users', employeeId));
    const blockOverride = userSnap.exists() ? (userSnap.data().blockOverride ?? false) : false;
    const isBlocked = !blockOverride && totalHours >= SHIFT_RULES.MAX_WEEKLY_HOURS;
    await updateDoc(doc(db, 'users', employeeId), { weeklyHours: totalHours, isBlocked });
  } catch {
    // sync failure is non-critical — approval already saved
  }
}

export async function approveAllPendingForEmployee(employeeId: string): Promise<number> {
  const q = query(
    collection(db, 'shifts'),
    where('employeeId', '==', employeeId),
    where('status', '==', 'pending')
  );
  const snap = await getDocs(q);
  if (snap.empty) return 0;
  await Promise.all(snap.docs.map(d => updateDoc(d.ref, { status: 'approved' })));
  const firstDate = (snap.docs[0].data() as Shift).date;
  await syncEmployeeWeeklyHours(employeeId, firstDate);
  return snap.docs.length;
}

export async function getWeeklyHours(employeeId: string, weekDate: Date): Promise<number> {
  const weekStart = getWeekStart(weekDate);
  const weekEnd = getWeekEnd(weekDate);
  const q = query(
    collection(db, 'shifts'),
    where('employeeId', '==', employeeId),
    where('status', '==', 'approved'),
    where('date', '>=', weekStart),
    where('date', '<=', weekEnd)
  );
  const snap = await getDocs(q);
  return snap.docs
    .map(d => d.data() as Shift)
    .reduce((sum, s) => sum + s.durationHours, 0);
}

export function validateShiftRequest(
  shiftType: ShiftType,
  customHours: number | null,
  lastShiftEnd: string | null,
  newShiftStart: string
): string | null {
  const hours = customHours ?? SHIFT_CONFIG[shiftType].hours;
  if (hours < SHIFT_RULES.MIN_SHIFT_HOURS) return `משמרת מינימום ${SHIFT_RULES.MIN_SHIFT_HOURS} שעות`;
  if (hours > SHIFT_RULES.MAX_SHIFT_HOURS) return `משמרת מקסימום ${SHIFT_RULES.MAX_SHIFT_HOURS} שעות`;

  if (lastShiftEnd) {
    const last = new Date(lastShiftEnd).getTime();
    const next = new Date(newShiftStart).getTime();
    const restHours = (next - last) / (1000 * 60 * 60);
    if (restHours < SHIFT_RULES.MIN_REST_HOURS)
      return `חובה ${SHIFT_RULES.MIN_REST_HOURS} שעות מנוחה בין משמרות`;
  }
  return null;
}

export async function requestShift(
  employeeId: string,
  employeeName: string,
  date: string,
  type: ShiftType,
  weeklyHours: number
): Promise<{ shift: Shift | null; error: string | null }> {
  if (!auth.currentUser || auth.currentUser.uid !== employeeId) {
    return { shift: null, error: 'לא מורשה לבצע פעולה זו' };
  }
  const day = new Date().getDay();
  if (day < 3 || day > 5) {
    return { shift: null, error: 'ניתן להגיש בקשות משמרת רק ביום רביעי עד שישי' };
  }
  const config = SHIFT_CONFIG[type];
  const newTotal = weeklyHours + config.hours;

  if (newTotal > SHIFT_RULES.MAX_WEEKLY_HOURS) {
    return { shift: null, error: `חרגת ממכסת ${SHIFT_RULES.MAX_WEEKLY_HOURS} שעות שבועיות` };
  }

  const validationError = validateShiftRequest(type, null, null, `${date}T${config.start}`);
  if (validationError) return { shift: null, error: validationError };

  const dupQ = query(
    collection(db, 'shifts'),
    where('employeeId', '==', employeeId),
    where('date', '==', date),
    where('type', '==', type),
    where('status', 'in', ['pending', 'approved'])
  );
  const dupSnap = await getDocs(dupQ);
  if (!dupSnap.empty) {
    return { shift: null, error: 'כבר קיימת בקשה לאותה משמרת בתאריך זה' };
  }

  const shift: Omit<Shift, 'id'> = {
    employeeId,
    employeeName,
    date,
    type,
    startTime: config.start,
    endTime: config.end,
    durationHours: config.hours,
    status: 'pending',
    createdAt: new Date().toISOString(),
  };

  const ref = await addDoc(collection(db, 'shifts'), shift);
  return { shift: { ...shift, id: ref.id }, error: null };
}

async function countApprovedForShift(date: string, type: ShiftType, excludeShiftId?: string): Promise<number> {
  const q = query(
    collection(db, 'shifts'),
    where('date', '==', date),
    where('type', '==', type),
    where('status', '==', 'approved')
  );
  const snap = await getDocs(q);
  return snap.docs.filter(d => d.id !== excludeShiftId).length;
}

export async function updateShiftStatus(
  shiftId: string,
  status: ShiftStatus,
  employeeId?: string,
  shiftDate?: string,
  shiftType?: ShiftType
): Promise<{ error?: string }> {
  if (status === 'approved' && shiftDate && shiftType) {
    const max = SHIFT_RULES.MAX_WORKERS_PER_SHIFT[shiftType];
    const current = await countApprovedForShift(shiftDate, shiftType, shiftId);
    if (current >= max) {
      return { error: `משמרת ${shiftType === 'morning' ? 'בוקר' : shiftType === 'afternoon' ? 'צהוריים' : 'לילה'} בתאריך ${shiftDate} מלאה (מקסימום ${max} עובדים)` };
    }
  }
  await updateDoc(doc(db, 'shifts', shiftId), { status });
  if (employeeId && shiftDate) {
    const dateLabel = shiftDate;
    const typeLabel = shiftType ? (SHIFT_CONFIG[shiftType] ? shiftType : '') : '';
    if (status === 'approved') {
      await syncEmployeeWeeklyHours(employeeId, shiftDate);
      await createNotification(
        employeeId,
        'shift_approved',
        'המשמרת שלך אושרה ✅',
        `משמרת ${typeLabel} בתאריך ${dateLabel} אושרה על ידי המנהל`
      );
    } else if (status === 'rejected') {
      await createNotification(
        employeeId,
        'shift_rejected',
        'המשמרת שלך נדחתה ❌',
        `משמרת ${typeLabel} בתאריך ${dateLabel} נדחתה על ידי המנהל`
      );
    }
  }
  return {};
}

export async function deleteShift(shiftId: string): Promise<void> {
  await deleteDoc(doc(db, 'shifts', shiftId));
}

export async function getShiftsByDate(date: string): Promise<Shift[]> {
  const q = query(collection(db, 'shifts'), where('date', '==', date));
  const snap = await getDocs(q);
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as Shift))
    .sort((a, b) => a.startTime.localeCompare(b.startTime));
}

export async function getEmployeeShifts(employeeId: string): Promise<Shift[]> {
  const q = query(
    collection(db, 'shifts'),
    where('employeeId', '==', employeeId)
  );
  const snap = await getDocs(q);
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as Shift))
    .sort((a, b) => b.date.localeCompare(a.date));
}

export async function getShiftManagerShifts(shiftManagerId: string): Promise<Shift[]> {
  const { getMyEmployees } = await import('./userService');
  const employees = await getMyEmployees(shiftManagerId);
  if (employees.length === 0) return [];
  const ids = employees.slice(0, 30).map(e => e.id);
  const q = query(collection(db, 'shifts'), where('employeeId', 'in', ids));
  const snap = await getDocs(q);
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as Shift))
    .sort((a, b) => b.date.localeCompare(a.date));
}

export async function getAllShifts(): Promise<Shift[]> {
  const snap = await getDocs(collection(db, 'shifts'));
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as Shift))
    .sort((a, b) => b.date.localeCompare(a.date));
}

export async function getAllPendingShifts(): Promise<Shift[]> {
  const q = query(collection(db, 'shifts'), where('status', '==', 'pending'));
  const snap = await getDocs(q);
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as Shift))
    .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
}

export async function getPendingShiftsByEmployees(employeeIds: string[]): Promise<Shift[]> {
  if (employeeIds.length === 0) return [];
  // Firestore מגביל 'in' ל-30 ערכים — מספיק למחלקה רגילה
  const ids = employeeIds.slice(0, 30);
  const q = query(
    collection(db, 'shifts'),
    where('status', '==', 'pending'),
    where('employeeId', 'in', ids)
  );
  const snap = await getDocs(q);
  return snap.docs
    .map(d => ({ id: d.id, ...d.data() } as Shift))
    .sort((a, b) => b.createdAt.localeCompare(a.createdAt));
}

export async function updateShiftHoursAndApprove(
  shiftId: string,
  startTime: string,
  endTime: string,
  durationHours: number,
  employeeId: string,
  shiftDate: string,
  originalHours?: number,
  shiftType?: ShiftType
): Promise<{ error?: string }> {
  if (shiftType) {
    const max = SHIFT_RULES.MAX_WORKERS_PER_SHIFT[shiftType];
    const current = await countApprovedForShift(shiftDate, shiftType, shiftId);
    if (current >= max) {
      return { error: `משמרת ${shiftType === 'morning' ? 'בוקר' : shiftType === 'afternoon' ? 'צהוריים' : 'לילה'} בתאריך ${shiftDate} מלאה (מקסימום ${max} עובדים)` };
    }
  }
  await updateDoc(doc(db, 'shifts', shiftId), {
    startTime,
    endTime,
    durationHours,
    status: 'approved',
  });
  await syncEmployeeWeeklyHours(employeeId, shiftDate);

  const typeLabel = shiftType ?? '';
  const hoursChanged = originalHours !== undefined && originalHours !== durationHours;
  const body = hoursChanged
    ? `משמרת ${typeLabel} בתאריך ${shiftDate} אושרה — שעות עודכנו מ-${originalHours} ל-${durationHours} שעות (${startTime}–${endTime})`
    : `משמרת ${typeLabel} בתאריך ${shiftDate} אושרה — ${startTime}–${endTime} (${durationHours} שעות)`;

  await createNotification(employeeId, 'shift_approved', 'המשמרת שלך אושרה ✅', body);
  return {};
}

File: backend/services/shift_service.py
from datetime import datetime, timedelta
from typing import Optional
from firebase_admin import firestore
from firebase_admin_init import get_db
from models import (
    SHIFT_CONFIG, MAX_WEEKLY_HOURS, MIN_SHIFT_HOURS,
    MAX_SHIFT_HOURS, MIN_REST_HOURS,
)


def _week_bounds(date_str: str) -> tuple[str, str]:
    # שבוע עברי: ראשון (0) עד שבת (6)
    d = datetime.strptime(date_str, "%Y-%m-%d")
    days_since_sunday = (d.weekday() + 1) % 7  # ראשון = 0
    week_start = d - timedelta(days=days_since_sunday)
    week_end = week_start + timedelta(days=6)
    return week_start.strftime("%Y-%m-%d"), week_end.strftime("%Y-%m-%d")


def get_weekly_hours(employee_id: str, date_str: str) -> float:
    db = get_db()
    week_start, week_end = _week_bounds(date_str)
    docs = (
        db.collection("shifts")
        .where("employeeId", "==", employee_id)
        .where("status", "==", "approved")
        .where("date", ">=", week_start)
        .where("date", "<=", week_end)
        .stream()
    )
    return sum(d.to_dict().get("durationHours", 0) for d in docs)


def _shift_end_dt(data: dict) -> datetime:
    """מחשב datetime מדויק של סוף משמרת, כולל לילה שעובר חצות."""
    base = datetime.strptime(data["date"], "%Y-%m-%d")
    end_hour, end_min = map(int, data["endTime"].split(":"))
    end_dt = base.replace(hour=end_hour, minute=end_min)
    if data["type"] == "night":
        end_dt += timedelta(days=1)
    elif data["type"] == "afternoon":
        # צהוריים נגמר ב-23:00 — נשאר אותו יום, אין שינוי
        pass
    return end_dt


def get_last_shift_end(employee_id: str, date_str: str, shift_type: str) -> Optional[datetime]:
    """מחזיר datetime של סוף המשמרת הקודמת הרלוונטית.

    בודק גם משמרות באותו יום (לצורך בוקר+לילה) וגם יום קודם (לצורך צהוריים+בוקר).
    """
    db = get_db()
    target_date = datetime.strptime(date_str, "%Y-%m-%d")
    prev_date = (target_date - timedelta(days=1)).strftime("%Y-%m-%d")

    # חפש במסגרת של יומיים אחורה
    docs = (
        db.collection("shifts")
        .where("employeeId", "==", employee_id)
        .where("status", "==", "approved")
        .where("date", ">=", prev_date)
        .where("date", "<=", date_str)
        .stream()
    )

    candidates = []
    for doc in docs:
        data = doc.to_dict()
        # דלג על משמרות באותו יום עם אותו סוג
        if data["date"] == date_str and data["type"] == shift_type:
            continue
        end_dt = _shift_end_dt(data)
        candidates.append(end_dt)

    return max(candidates) if candidates else None


def validate_shift(
    shift_type: str,
    date_str: str,
    duration_hours: float,
    employee_id: str,
) -> Optional[str]:
    if duration_hours < MIN_SHIFT_HOURS:
        return f"משמרת מינימום {MIN_SHIFT_HOURS} שעות"
    if duration_hours > MAX_SHIFT_HOURS:
        return f"משמרת מקסימום {MAX_SHIFT_HOURS} שעות"

    config = SHIFT_CONFIG[shift_type]
    start_hour, start_min = map(int, config["start"].split(":"))
    shift_start_dt = datetime.strptime(date_str, "%Y-%m-%d").replace(
        hour=start_hour, minute=start_min
    )

    last_end = get_last_shift_end(employee_id, date_str, shift_type)
    if last_end:
        rest_hours = (shift_start_dt - last_end).total_seconds() / 3600
        if rest_hours < MIN_REST_HOURS:
            return f"חובה {MIN_REST_HOURS} שעות מנוחה בין משמרות (נותרו {MIN_REST_HOURS - rest_hours:.1f} שעות)"

    return None


def request_shift(
    employee_id: str,
    employee_name: str,
    date_str: str,
    shift_type: str,
    custom_hours: Optional[float] = None,
) -> dict:
    db = get_db()
    config = SHIFT_CONFIG[shift_type]
    duration = custom_hours if custom_hours is not None else config["hours"]

    # Check weekly hours
    weekly_hours = get_weekly_hours(employee_id, date_str)
    if weekly_hours + duration > MAX_WEEKLY_HOURS:
        return {"error": f"חרגת ממכסת {MAX_WEEKLY_HOURS} שעות שבועיות ({weekly_hours:.0f} + {duration:.0f} = {weekly_hours + duration:.0f})"}

    # Validate rest and duration
    error = validate_shift(shift_type, date_str, duration, employee_id)
    if error:
        return {"error": error}

    shift = {
        "employeeId": employee_id,
        "employeeName": employee_name,
        "date": date_str,
        "type": shift_type,
        "startTime": config["start"],
        "endTime": config["end"],
        "durationHours": duration,
        "status": "pending",
        "createdAt": datetime.utcnow().isoformat(),
    }
    ref = db.collection("shifts").add(shift)
    return {"id": ref[1].id, **shift}


def approve_shift(shift_id: str) -> None:
    db = get_db()
    shift_ref = db.collection("shifts").document(shift_id)
    shift_data = shift_ref.get().to_dict()
    if not shift_data:
        raise ValueError("Shift not found")

    shift_ref.update({"status": "approved"})

    employee_id = shift_data["employeeId"]
    new_weekly = get_weekly_hours(employee_id, shift_data["date"])

    user_ref = db.collection("users").document(employee_id)
    is_blocked = new_weekly >= MAX_WEEKLY_HOURS
    user_ref.update({"weeklyHours": new_weekly, "isBlocked": is_blocked})

    if is_blocked:
        _notify_manager_hours_limit(employee_id, shift_data["employeeName"], new_weekly)

    _notify_employee(employee_id, "shift_approved",
                     "המשמרת שלך אושרה",
                     f"משמרת {shift_data['date']} ({shift_data['type']}) אושרה על ידי המנהל")


def reject_shift(shift_id: str, employee_id: str, employee_name: str, date_str: str, shift_type: str) -> None:
    db = get_db()
    db.collection("shifts").document(shift_id).update({"status": "rejected"})
    _notify_employee(employee_id, "shift_rejected",
                     "המשמרת שלך נדחתה",
                     f"משמרת {date_str} ({shift_type}) נדחתה על ידי המנהל")


def update_shift_hours(shift_id: str, start_time: str, end_time: str, duration_hours: float) -> None:
    db = get_db()
    db.collection("shifts").document(shift_id).update({
        "startTime": start_time,
        "endTime": end_time,
        "durationHours": duration_hours,
    })


def _notify_employee(employee_id: str, notif_type: str, title: str, body: str) -> None:
    db = get_db()
    db.collection("notifications").add({
        "userId": employee_id,
        "type": notif_type,
        "title": title,
        "body": body,
        "read": False,
        "createdAt": datetime.utcnow().isoformat(),
    })


def _notify_manager_hours_limit(employee_id: str, employee_name: str, hours: float) -> None:
    db = get_db()
    managers = db.collection("users").where("role", "==", "manager").stream()
    for mgr in managers:
        db.collection("notifications").add({
            "userId": mgr.id,
            "type": "hours_limit",
            "title": f"עובד הגיע לגבול השעות",
            "body": f"{employee_name} הגיע ל-{hours:.0f} שעות שבועיות (מקסימום {MAX_WEEKLY_HOURS}). ניתן לבטל חסימה.",
            "read": False,
            "createdAt": datetime.utcnow().isoformat(),
        })
